core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -611,7 +613,17 @@ private void applyInitializingShard(final ClusterState state, IndexService index
         }
 
         indexShard.startRecovery(nodes.getLocalNode(), sourceNode, recoveryTargetService,
-            new RecoveryListener(shardRouting, indexService), repositoriesService);
+            new RecoveryListener(shardRouting, indexService), repositoriesService, (type, mapping) -> {
+                try {
+                    nodeServicesProvider.getClient().admin().indices().preparePutMapping()
```

Comment:
> The IndexShard can just verify that the source and target mapping are identical, and if not fail the allocation? 

it can't it has to be the master.

> this is tricky - how do we secure this for example? 

this is the ideal situation, even if you keep on writing to the index at the time we copy the mapping we already got a snapshot of the shard and the docs that are in the snapshot are guaranteed in the mapping. You are prone to races if you copy ahead of time and you are prone to races if you check. I am sorry your suggestion make it error prone and more complex. ES is inherent concurrent which makes anything hard. Unless we fix that we can't make such assumptions.